### PR TITLE
Fix spell book status on hovering

### PR DESCRIPTION
**Pull request by vincent-grosbois**
This contains a small refactor of spellbook (1st commit) and the fix for issue #51

### CODE REVIEW

File: src/fheroes2/spell/spell_book.cpp [hunk 1]
```diff
@@ -101,12 +98,12 @@ Spell SpellBook::Open( const HeroBase & hero, int filt, bool canselect ) const
     const fheroes2::Rect cmbt_rt( pos.x + bookmarkCombatoOffset.x, pos.y + bookmarkCombatoOffset.y, bookmark_cmbt.width(), bookmark_cmbt.height() );
     const fheroes2::Rect clos_rt( pos.x + bookmarkCloseOffset.x, pos.y + bookmarkCloseOffset.y, bookmark_clos.width(), bookmark_clos.height() );
 
-    Spell curspell( Spell::NONE );
+    Spell curspell = Spell::NONE;
```
Comment (ihhub) [inline src/fheroes2/spell/spell_book.cpp:101]: Could you please explain why we changed from a parametrized constructor to assignment constructor here?

File: src/fheroes2/resource/artifact.cpp [hunk 2]
```diff
@@ -1006,8 +1008,15 @@ bool ArtifactsBar::ActionBarLeftMouseDoubleClick( Artifact & art )
     if ( art() == Artifact::MAGIC_BOOK ) {
         if ( can_change )
             const_cast<Heroes *>( hero )->EditSpellBook();
-        else
-            hero->OpenSpellBook( SpellBook::ALL, false );
+        else {
+            std::function<void( const std::string & )> statusCallback = [this]( const std::string & status ) {
```
Comment (ihhub) [inline src/fheroes2/resource/artifact.cpp:1012]: Let's pass only `&bar` of the current class. We don't need the whole class to be exposed. Nothing wrong in the current implementation, just to follow incapsulation principle :)

File: src/fheroes2/spell/spell_book.cpp [hunk 3]
```diff
@@ -33,32 +34,39 @@
...
+    const int SpellsPerPage = 6;
+
     const fheroes2::Point bookmarkInfoOffset( 123, 273 );
     const fheroes2::Point bookmarkAdvOffset( 266, 269 );
     const fheroes2::Point bookmarkCombatoOffset( 299, 276 );
     const fheroes2::Point bookmarkCloseOffset( 416, 280 );
+
+    bool SpellBookSortingSpell( const Spell & spell1, const Spell & spell2 )
+    {
+        return std::string( spell1.GetName() ) < std::string( spell2.GetName() );
```
Comment (ihhub) [inline src/fheroes2/spell/spell_book.cpp:48]: Please compare it using `strcmp` without **std::string** usage.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vincent-grosbois , I left multiple comment here. Could you please take a look?

File: src/fheroes2/spell/spell_book.cpp (hunk 1 above)
Comment (vincent-grosbois) [reply src/fheroes2/spell/spell_book.cpp:101]: I don't think it's assignment constructor in that case, it should call the same constructor.
I thought this is more readable

File: src/fheroes2/resource/artifact.h [hunk 4]
```diff
@@ -232,10 +232,11 @@ class BagArtifacts : public std::vector<Artifact>
     std::string String( void ) const;
 };
 
+class StatusBar;
```
Comment (ihhub) [inline src/fheroes2/resource/artifact.h:235]: Please forward declaration of this class at the top of the file.

File: src/fheroes2/heroes/heroes_base.h [hunk 5]
```diff
@@ -112,7 +112,7 @@ class HeroBase : public Skill::Primary, public MapPosition, public BitModes, pub
 
     std::vector<Spell> GetSpells( int lvl = -1 ) const;
     void EditSpellBook( void );
-    Spell OpenSpellBook( int filter, bool ) const;
+    Spell OpenSpellBook( SpellBook::Filter filter, bool canSelect, std::function<void( const std::string & )> * statusCallback = nullptr ) const;
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_base.h:115]: Please remove default parameter and just explicitly set it in one call.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vincent-grosbois , I left few comments only. Could you please take a look?

File: src/fheroes2/resource/artifact.cpp [hunk 6]
```diff
@@ -1006,8 +1008,16 @@ bool ArtifactsBar::ActionBarLeftMouseDoubleClick( Artifact & art )
     if ( art() == Artifact::MAGIC_BOOK ) {
         if ( can_change )
             const_cast<Heroes *>( hero )->EditSpellBook();
-        else
-            hero->OpenSpellBook( SpellBook::ALL, false );
+        else {
+            StatusBar * const statusBar = this->_statusBar;
+            std::function<void( const std::string & )> statusCallback = [statusBar]( const std::string & status ) {
```
[the remark this answers is not in the document]
Comment (vincent-grosbois) [reply src/fheroes2/resource/artifact.cpp:1013]: > 
> 
> We can't we pass `[&_statusBar]` without extra variable?

no, this is a C++17 feature apparently
C++11 it's either this or the initial version, ie using this in the closure

**Comment by vincent-grosbois** [comment]
@ihhub  done!

**Comment by ihhub** [comment]
@vincent-grosbois , thank you so much for adding this feature!

**Pull request merged**